mantidproject/mantid
# Lagrange: enable nexus data processing

This is a PR to include NeXus in the Lagrange workflow. It implements a new loader: `LoadILLLagrange` that handles loading of Lagrange data from NeXus, including metadata, and setting `Ei` scanned variable into metadata. This part includes adding a mock instrument with arbitrary dimensions and distances. It is however imperative to have this instrument defined to be able to set spectra as monitor or data, and later on to use an Instrument Parameter File (not implemented as of now).

Then, loading of NeXus files is added to the reduction workflow via `LoadAndMerge` algorithm. In case of multiple files, data is concatenated onto a single X axis using `Ei` values from sample logs as X values. Following that, the X axis is sorted so the merging algorithm can properly check if there are no scans placed too close in energy. The rest of the reduction is the same as for ASCII data. A number of unit tests checking the validity of processing has been added.

**To test:**

To test:

1. Code review.
2. The new tests should be passing. The test is based on the future Lagrange NeXus , and a bunch of synthetic data based on what was provided by data acquisition team at ILL.
3. Process the following file, by setting it to SampleRuns: [too_close_scans.zip](https://github.com/mantidproject/mantid/files/10720667/too_close_scans.zip), make sure that there are no bin centres closer to each other than 1e-2 meV.
4. Process the following files: [014412_files.zip](https://github.com/mantidproject/mantid/files/10720678/014412_files.zip),
 make sure that the X axis has values going from 35 to 65 meV.

Fixes #35152. 


Release notes and documentation updated.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
